qcodes/utils/plotting.py

```diff
@@ -42,4 +67,180 @@ def auto_range_iqr(data_array: np.ndarray) -> Tuple[float, float]:
     else:
         vmin = max(q1 - 1.5*IQR, zmin)
         vmax = min(q3 + 1.5*IQR, zmax)
+        # do not clipp more than cutoff_percentile:
+        vmin = min(vmin, pmin)
+        vmax = max(vmax, pmax)
     return vmin, vmax
+
+
+# Matplotlib functions
+
+DEFAULT_COLOR_OVER = 'Magenta'
+DEFAULT_COLOR_UNDER = 'Cyan'
+
+
+def apply_color_scale_limits(colorbar: matplotlib.pyplot.colorbar,
+                             new_lim: Tuple[Optional[float], Optional[float]],
+                             data_lim: Optional[Tuple[float, float]]=None,
+                             data_array: Optional[np.ndarray]=None,
+                             color_over: Optional[Any]=DEFAULT_COLOR_OVER,
+                             color_under: Optional[Any]=DEFAULT_COLOR_UNDER
+                             ) -> None:
+    """
+    Applies limits to colorscale and updates extend.
+
+    This function applies the limits `new_lim` to the heatmap plot associated
+    with the provided `colorbar`, updates the colorbar limits, and also adds
+    the colorbar clipping indicators in form of small triangles on the top and
+    bottom of the colorbar, according to the where the limits are exceeded.
+
+    Args:
+        colorbar: The actual colorbar to be updated
+        new_lim: 2-tuple of the desired minimum and maximum value of the color
+            scale. If any is `None` it will be left unchanged.
+        data_lim: 2-tuple of the actual minimum and maximum value of the data.
+            If left out the minimum and maximum are deduced from the provided
+            data, or the data associated with the colorbar.
+        data_array: numpy array containing the data to be considered for
+            scaling. Must be left out if `data_lim` is provided. If neither is
+            provided the data associated with the colorbar is used.
+        color_over: Matplotlib color representing the datapoints clipped by the
+            upper limit
+        color_under: Matplotlib color representing the datapoints clipped by
+            the lower limit
+    """
+    # browse the input data and make sure that `data_lim` and `new_lim` are
+    # available
+    if not isinstance(colorbar.mappable, matplotlib.collections.QuadMesh):
+        raise RuntimeError('Can only scale mesh data, but received '
+                           f'"{type(colorbar.mappable)}" instead')
+    if data_lim is None:
+        if data_array is None:
+            data_array = colorbar.mappable.get_array()
+        data_lim = np.nanmin(data_array), np.nanmax(data_array)
+    else:
+        if data_array is not None:
+            raise RuntimeError('You cannot specify `data_lim` and `data_array` '
```

Comment:
could the message be rephrased a bit, so that it's clear that both cannot be supplised *at the same time*? also, a punctuation before "please" is missing.